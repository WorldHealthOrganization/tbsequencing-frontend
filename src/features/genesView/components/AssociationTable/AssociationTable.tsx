/** @jsxImportSource @emotion/react */

import React from 'react';
import { SerializedStyles } from '@emotion/react';
import AppPaper from '../../../../components/AppPaper';
import H3 from '../../../../components/typography/H3';
import { associationTableStyles } from './styles';
import AppButton from '../../../../components/AppButton';
import LoadingWrapper from '../../../../components/LoadingWrapper';

export interface IAssociationTableData {
  id: number;
  leftColValue: number | string;
  rightColValue: number | string;
}

interface IAssociationTableProps {
  leftColumnItemComponent?: React.FC<{
    item: IAssociationTableData, onClick?: (value: IAssociationTableData) => void
  }>;
  tableHeader?: string;
  data?: IAssociationTableData[];
  onClick?: (value: IAssociationTableData) => void;
  isLoading: boolean;
  containerStyles?: SerializedStyles | {}
  noDataText?: string;
}

interface ILeftColumnButtonProps {
  item: IAssociationTableData,
  onClick?: (value: IAssociationTableData) => void;
}

export const LeftColumnButton = ({ item, onClick }: ILeftColumnButtonProps) => (
  <AppButton
    style={associationTableStyles.leftColumn}
    size="small"
    onClick={() => typeof onClick === 'function' && onClick(item)}
    variant="text"
    underlined
  >
    {item.leftColValue}
  </AppButton>
);

const DefaultLeftColumnItem = ({ text }: { text?: string }) => (
  <H3 style={associationTableStyles.defaultColumn}>{text}</H3>
);

const AssociationTable = ({
  tableHeader,
  data,
  leftColumnItemComponent,
  onClick,
  isLoading,
  containerStyles = {},
  noDataText = 'No data',
}: IAssociationTableProps) => {
  const LeftColumnItem = leftColumnItemComponent || DefaultLeftColumnItem;
  return (
    <AppPaper style={{
      ...associationTableStyles.paper,
      ...containerStyles,
    }}
    >
      <div>
        {tableHeader && <H3>{tableHeader}</H3>}
      </div>
      <LoadingWrapper isLoading={isLoading} centered>
        <div css={associationTableStyles.itemsContainer}>
          {!data?.length ? noDataText
            : data.map((item) => (
              <div key={item.id} css={associationTableStyles.singlePair}>
                <LeftColumnItem item={item} onClick={onClick} />
                <H3 style={associationTableStyles.rightColumnText}>
                  {item.rightColValue}
                </H3>
              </div>
            ))}
        </div>
      </LoadingWrapper>
    </AppPaper>
  );
};

export default AssociationTable;
