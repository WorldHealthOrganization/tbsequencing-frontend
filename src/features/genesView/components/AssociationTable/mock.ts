import { IGenesOverview, ISearchCount } from '../../models';

export const mostSearches: ISearchCount[] = [
  {
    counter: 112, geneName: 'rpoB', geneDbCrossrefId: 114, locusTag: 'aboba',
  },
  {
    counter: 97, geneName: 'rpoC', geneDbCrossrefId: 154, locusTag: 'aboba',
  },
  {
    counter: 90, geneName: 'Rv0669c', geneDbCrossrefId: 144, locusTag: '',
  },
  {
    counter: 50, geneName: 'J8670_RS10355', geneDbCrossrefId: 116, locusTag: '',
  },
  {
    counter: 22, geneName: 'J8670_RS10350', geneDbCrossrefId: 118, locusTag: '',
  },
  {
    counter: 18, geneName: 'cfp21', geneDbCrossrefId: 125, locusTag: '',
  },
  {
    counter: 12, geneName: 'J8670_RS10345', geneDbCrossrefId: 119, locusTag: '',
  },
  {
    counter: 9, geneName: 'vapB8', geneDbCrossrefId: 120, locusTag: '',
  },
  {
    counter: 3, geneName: 'vapC8', geneDbCrossrefId: 121, locusTag: '',
  },
];

export const simpleData: IGenesOverview = {
  endPos: 763325,
  id: 694,
  geneDbCrossrefId: 694,
  geneDescription: 'DNA-directed RNA polymerase subunit beta',
  geneName: 'rpoB',
  geneType: 'NP_215181.1',
  locusTag: 'Rv0667',
  ncbiId: '888164',
  proteinLength: 1172,
  startPos: 759807,
  strand: 1,
};

export const drugsTable = [
  { code: 'RIF', drugName: 'Rifampicine', drugId: 2 },
  { code: 'INH', drugName: 'Isoniazid', drugId: 1 },
  { code: 'EMB', drugName: 'Ethambutol', drugId: 4 },
  { code: 'LZD', drugName: 'Linezolid', drugId: 21 },
];

export const association = [
  { geneName: 'cfp21', drugs: [{ code: 'INH' }, { code: 'ZLD' }], geneDbCrossref: 1 },
  { geneName: 'J8670 RS10345', drugs: [{ code: 'INH' }, { code: 'ZLD' }], geneDbCrossref: 2 },
  { geneName: 'rpoB', drugs: [{ code: 'INH' }, { code: 'ZLD' }], geneDbCrossref: 694 },
  { geneName: 'Rv2490a', drugs: [{ code: 'INH' }, { code: 'ZLD' }], geneDbCrossref: 7 },
];
