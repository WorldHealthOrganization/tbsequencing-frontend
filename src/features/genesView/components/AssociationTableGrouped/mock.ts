export const association = [
  {
    geneName: 'cfp21', geneDbCrossrefId: 1525, drugId: 1, tier: 1,
  },
  {
    geneName: 'J8670 RS10345', geneDbCrossrefId: 1526, drugId: 1, tier: 1,
  },
  {
    geneName: 'rpoB', geneDbCrossrefId: 1527, drugId: 1, tier: 2,
  },
  {
    geneName: 'Rv2490a', geneDbCrossrefId: 1526, drugId: 1, tier: 2,
  },
];
