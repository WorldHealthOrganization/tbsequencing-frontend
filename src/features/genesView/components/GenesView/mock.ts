export const mockGenes = [
  { geneId: 1, geneName: 'rpoC' },
  { geneId: 2, geneName: 'rpoB' },
  { geneId: 3, geneName: 'rpoA' },
  { geneId: 4, geneName: 'vpo8B' },
  { geneId: 5, geneName: 'vpo8A' },
  { geneId: 6, geneName: 'Rv0669c' },
  { geneId: 7, geneName: 'cfp21' },
  { geneId: 8, geneName: 'J8670_RS10350' },
  { geneId: 9, geneName: 'J8670_RS10355' },
];
