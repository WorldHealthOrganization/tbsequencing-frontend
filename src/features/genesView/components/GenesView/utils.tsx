/** @jsxImportSource @emotion/react */
import React from 'react';
import { formatDistance, parseISO } from 'date-fns';
import { IGenesOverview, ISearchCount, ISearchRecent } from '../../models';
import { GeneAssociationData, GeneAssociationGroupedData } from '../../../../services/drugsApi/models';
import { cdsLink, geneOverviewLink } from './styles';

const getDayByDate = (date: string) => (date.includes('hours') ? 'Today' : date);

export const formatRecentlySearch = (searchRecentlyData?: ISearchRecent[]) => {
  if (!searchRecentlyData) return [];

  return searchRecentlyData.map((item) => ({
    id: item.geneDbCrossrefId,
    leftColValue: item.geneName || item.locusTag,
    rightColValue: getDayByDate(
      `${formatDistance(parseISO(item.date as string), new Date(), { addSuffix: true })}`,
    ),
  }));
};

export const formatMostSearch = (searchMostData?: ISearchCount[]) => {
  if (!searchMostData) return [];

  return searchMostData.map((item) => ({
    id: item.geneDbCrossrefId,
    leftColValue: item.geneName || item.locusTag,
    rightColValue: item.counter,
  }));
};

export const formatDrugsTableGenesView = (data?: any[]) => {
  if (!data) return [];

  return data.map((item) => ({
    id: item.drugId,
    leftColValue: item.code,
    rightColValue: item.drugName,
  }));
};

export const formatResistantDataGrouped = (data?: GeneAssociationGroupedData[]) => {
  if (!data?.length) return [];

  return data.map((item) => ({
    id: item.geneDbCrossrefId,
    leftColValue: item.geneName || item.locusTag,
    tier: item.tier,
  }));
};

export const formatResistantData = (data?: GeneAssociationData[]) => {
  if (!data?.length) return [];

  return data.map((item) => ({
    id: item.geneDbCrossref,
    leftColValue: item.geneName,
    rightColValue: item.drugs.map((d) => d.code).join(', '),
  }));
};

export const getGenomeContextLength = (gene?: IGenesOverview) => {
  if (!gene) return {};
  const length = gene.endPos - gene.startPos;
  const startPos = parseInt(String(gene.startPos - length * 1.5), 10);
  const endPos = parseInt(String(gene.endPos + length * 1.5), 10);
  return {
    startPos: startPos > 0 ? startPos : 1,
    endPos,
  };
};

export const formatGenesOverviewData = (data?: IGenesOverview) => {
  if (!data) return [];
  const {
    geneName,
    locusTag,
    geneType,
    geneDescription,
    startPos,
    endPos,
    ncbiId,
    proteinLength,
  } = data;

  return [
    { title: 'Gene Symbol', value: geneName || locusTag },
    {
      title: 'NCBI Gene ID',
      value: ncbiId,
      element: (
        <a
          css={geneOverviewLink}
          target="_blank"
          href={`https://www.ncbi.nlm.nih.gov/gene/${ncbiId}`}
          rel="noreferrer"
        >
          {ncbiId}
        </a>
      ),
    },
    { title: 'Gene Description', value: geneDescription },
    { title: 'Locus Tag', value: locusTag },
    { title: 'Gene type', value: geneType ? 'protein coding' : null },
    {
      title: 'Location',
      value: endPos && startPos
        ? `${startPos.toLocaleString('en-US')}..${endPos.toLocaleString('en-US')}`
        : null,
    },
    {
      title: 'Length',
      value: endPos && startPos
        ? `${(endPos - startPos).toLocaleString('en-US')} nt`
        : null,
    },
    {
      title: 'CDS',
      value: geneType,
      element: (
        <a
          css={[geneOverviewLink, cdsLink]}
          target="_blank"
          href={`https://www.ncbi.nlm.nih.gov/protein/${geneType}`}
          rel="noreferrer"
        >
          {geneType}
        </a>
      ),
    },
    {
      title: 'Protein Length',
      value: proteinLength
        ? `${proteinLength.toLocaleString('en-US')} aa`
        : null,
    },
  ];
};
