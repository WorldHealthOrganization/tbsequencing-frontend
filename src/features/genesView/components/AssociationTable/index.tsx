import AssociationTable from './AssociationTable';

export { LeftColumnButton } from './AssociationTable';
export default AssociationTable;
