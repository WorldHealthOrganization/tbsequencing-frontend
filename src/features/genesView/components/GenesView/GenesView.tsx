/** @jsxImportSource @emotion/react */
import React, { SyntheticEvent, useEffect, useState } from 'react';
import { useSearchParams } from 'react-router-dom';
import BasePage from '../../../../components/BasePage';
import AppPaper from '../../../../components/AppPaper';
import AutocompleteInput from '../../../../components/AutocompleteInput';
import { genesViewStyles } from './styles';
import H3 from '../../../../components/typography/H3';
import {
  useGetMostSearchHistoryQuery,
  useGetRecentSearchHistoryQuery,
} from '../../../../services/genesApi/genesApi';
import { useAutocompleteQuery } from '../../hooks/useAutocompleteQuery';
import { useDebounce } from '../../../../hooks/useDebounce';
import GeneNotSelectedContent from './GeneNotSelectedContent';
import GeneSelectedContent from './GeneSelectedContent';
import { IGene } from '../../../../services/genesApi/models';
import { formatMostSearch, formatRecentlySearch } from './utils';

const GenesView = () => {
  const [searchParams] = useSearchParams();

  const defaultGene: IGene = {
    geneDbCrossrefId: Number(searchParams.get('geneDbCrossrefId')),
    geneName: searchParams.get('geneName'),
    locusTag: null,
    id: Number(searchParams.get('geneDbCrossrefId')),
    endPos: 1,
    startPos: 0,
    strand: -1,
  };

  const [inputValue, setInputValue] = useState('');
  const [
    selectedValue,
    setSelectedValue,
  ] = useState<IGene | null>(defaultGene.geneDbCrossrefId ? defaultGene : null);
  const debouncedValue = useDebounce<string>(inputValue, 500);

  const {
    data: mostSearch = [],
    isLoading: isMostSearchLoading, refetch: mostSearchHistoryRefetch,
  } = useGetMostSearchHistoryQuery(undefined, { refetchOnMountOrArgChange: true });

  const {
    data: recentSearch = [],
    isLoading: isRecentSearchLoading, refetch: resentSearchRefetch,
  } = useGetRecentSearchHistoryQuery(undefined, { refetchOnMountOrArgChange: true });

  const {
    data: genes = [], fetchData, isLoading,
  } = useAutocompleteQuery();

  const getOptions = () => {
    const inputEmpty = !inputValue;
    const mostSearchedData = mostSearch
      .slice(-10)
      .sort((prev, next) => next.counter - prev.counter);

    return inputEmpty ? mostSearchedData : genes;
  };

  const options = getOptions();

  const recentSearchLocal = formatRecentlySearch(recentSearch);
  const mostSearchLocal = formatMostSearch(mostSearch);

  const handleInputChange = (e: SyntheticEvent, value: string) => {
    setInputValue(value);
  };

  useEffect(() => {
    if (debouncedValue !== '') {
      fetchData({ search: debouncedValue });
    }
  }, [debouncedValue]);

  const patchedSelectedGene = genes.length === 1
    ? { ...selectedValue, ...genes[0] }
    : selectedValue;

  return (
    <BasePage pageHeader="Gene View">
      <AppPaper>
        <div css={genesViewStyles.autoCompleteContainer}>
          <AutocompleteInput
            isLoading={isLoading}
            onSelectionChanged={(e, value) => {
              if (value === null) {
                mostSearchHistoryRefetch();
                resentSearchRefetch();
              }

              setSelectedValue(value);
            }}
            selectedValue={selectedValue}
            onChange={handleInputChange}
            value={inputValue}
            options={options}
            style={genesViewStyles.sxAutoComplete}
          />
          <H3
            style={genesViewStyles.autoCompleteInputLabel}
          >
            Search Genes by Gene Symbol, NCBI ID or Locus Tag
          </H3>
        </div>
      </AppPaper>
      {
        !selectedValue
          ? (
            <GeneNotSelectedContent
              isMostSearchLoading={isMostSearchLoading}
              isRecentSearchLoading={isRecentSearchLoading}
              mostSearch={mostSearchLocal}
              recentSearch={recentSearchLocal}
              onGeneClick={(item) => {
                setSelectedValue({
                  endPos: 0,
                  startPos: 0,
                  strand: -1,
                  id: item.id,
                  geneDbCrossrefId: item.id,
                  geneName: item.leftColValue as string,
                  locusTag: item.leftColValue as string,
                });
              }}
            />
          )
          : <GeneSelectedContent selectedGene={(patchedSelectedGene) as IGene} />
      }
    </BasePage>
  );
};

export default GenesView;
