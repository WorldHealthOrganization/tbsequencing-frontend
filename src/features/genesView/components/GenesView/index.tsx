import GenesView from './GenesView';

export default GenesView;
