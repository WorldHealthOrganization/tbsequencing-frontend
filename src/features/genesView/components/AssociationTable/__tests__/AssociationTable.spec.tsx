import React from 'react';
import { fireEvent } from '@testing-library/react';
import { getTreeJSONWithRedux, render, screen } from '../../../../../utils/testUtils/jestUtils';
import AssociationTable, { LeftColumnButton } from '../AssociationTable';
import { formatResistantData } from '../../GenesView/utils';
import { association } from '../mock';
import { GeneAssociationData } from '../../../../../services/drugsApi/models';

const mockData = formatResistantData(association as GeneAssociationData[]);

describe('AssociationTable', () => {
  const mockOnClick = jest.fn();
  test('renders AssociationTable component without error', () => {
    render(
      <AssociationTable
        tableHeader="Table header"
        data={mockData}
        isLoading={false}
      />,
    );
  });

  test('calls onClick function on click', () => {
    render(
      <AssociationTable
        tableHeader="Table header"
        data={mockData}
        leftColumnItemComponent={LeftColumnButton}
        onClick={mockOnClick}
        isLoading={false}
      />,
    );

    const clickItem = screen.getByText('rpoB');

    expect(clickItem).toBeInTheDocument();

    fireEvent.click(clickItem);

    expect(mockOnClick).toHaveBeenCalled();
  });

  test('Should render correctly', () => {
    const tree = getTreeJSONWithRedux(
      <AssociationTable tableHeader="Table header" data={mockData} isLoading={false} />,
    );

    expect(tree).toMatchSnapshot();
  });
});
