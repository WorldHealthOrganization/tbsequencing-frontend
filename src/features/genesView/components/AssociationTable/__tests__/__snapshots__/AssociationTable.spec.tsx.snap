// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssociationTable Should render correctly 1`] = `
<div
  className="css-9j8ib4-AppPaper"
>
  <div>
    <p
      className="MuiTypography-root MuiTypography-body1 css-5th3yp-MuiTypography-root-BaseText"
    >
      Table header
    </p>
  </div>
  <div
    className="css-hxjvcr-AssociationTable"
  >
    <div
      className="css-1pwa69v-AssociationTable"
    >
      <p
        className="MuiTypography-root MuiTypography-body1 css-1h9ij6s-MuiTypography-root-BaseText"
      />
      <p
        className="MuiTypography-root MuiTypography-body1 css-1h9ij6s-MuiTypography-root-BaseText"
      >
        INH, ZLD
      </p>
    </div>
    <div
      className="css-1pwa69v-AssociationTable"
    >
      <p
        className="MuiTypography-root MuiTypography-body1 css-1h9ij6s-MuiTypography-root-BaseText"
      />
      <p
        className="MuiTypography-root MuiTypography-body1 css-1h9ij6s-MuiTypography-root-BaseText"
      >
        INH, ZLD
      </p>
    </div>
    <div
      className="css-1pwa69v-AssociationTable"
    >
      <p
        className="MuiTypography-root MuiTypography-body1 css-1h9ij6s-MuiTypography-root-BaseText"
      />
      <p
        className="MuiTypography-root MuiTypography-body1 css-1h9ij6s-MuiTypography-root-BaseText"
      >
        INH, ZLD
      </p>
    </div>
    <div
      className="css-1pwa69v-AssociationTable"
    >
      <p
        className="MuiTypography-root MuiTypography-body1 css-1h9ij6s-MuiTypography-root-BaseText"
      />
      <p
        className="MuiTypography-root MuiTypography-body1 css-1h9ij6s-MuiTypography-root-BaseText"
      >
        INH, ZLD
      </p>
    </div>
  </div>
</div>
`;
